numpy/core/_exceptions.py

```diff
@@ -122,21 +122,89 @@ class TooHardError(RuntimeError):
 
 @set_module('numpy')
 class AxisError(ValueError, IndexError):
-    """ Axis supplied was invalid. """
+    """Axis supplied was invalid.
+
+    Raised whenever an ``axis`` parameter is specified that is larger than
+    the number of array dimensions.
+
+    .. versionadded:: 1.22
+
+        Added the ``axis`` and ``ndim`` attributes.
+
+    Parameters
+    ----------
+    axis : int or str
+        The out of bounds axis or a custom exception message.
+        If an axis is provided, then `ndim` should be specified as well.
+    ndim : int, optional
+        The number of array dimensions.
+    msg_prefix : str, optional
+        A prefix for the exception message.
+
+    Attributes
+    ----------
+    axis : int, optional
+        The out of bounds axis or ``None`` if a custom exception
+        message was provided.
+    ndim : int, optional
+        The number of array dimensions or ``None`` if a custom exception
+        message was provided.
+
+    Examples
+    --------
+    >>> array_1d = np.arange(10)
+    >>> np.cumsum(array_1d, axis=1)
+    Traceback (most recent call last):
+      ...
+    numpy.AxisError: axis 1 is out of bounds for array of dimension 1
+
+    The class constructor generally takes the axis and arrays'
+    dimensionality as arguments:
+
+    >>> np.AxisError(2, 1, prefix='error')
+    numpy.AxisError('error: axis 2 is out of bounds for array of dimension 1')
+
+    Alternativelly, a custom exception message can be passed:
+
+    >>> np.AxisError('Custom error message')
+    numpy.AxisError('Custom error message')
+
+    """
+
+    __slots__ = ("axis", "ndim")
+
     def __init__(self, axis, ndim=None, msg_prefix=None):
         # single-argument form just delegates to base class
         if ndim is None and msg_prefix is None:
             msg = axis
+            self.axis = None
+            self.ndim = None
 
         # do the string formatting here, to save work in the C code
         else:
             msg = ("axis {} is out of bounds for array of dimension {}"
                    .format(axis, ndim))
             if msg_prefix is not None:
                 msg = "{}: {}".format(msg_prefix, msg)
+            self.axis = axis
+            self.ndim = ndim
 
         super().__init__(msg)
 
+    @classmethod
+    def _reconstruct(cls, axis, ndim, args):
+        """Auxiliary instance constructor used by `__reduce__`."""
+        self = super().__new__(cls)
+        self.axis = axis
+        self.ndim = ndim
+        self.args = args
+        return self
+
+    def __reduce__(self):
+        """Return state information for `pickle`."""
+        cls = type(self)
+        return cls._reconstruct, (self.axis, self.ndim, self.args)
+
```

Comment:
Can you add a comment explaining why we need this? Is it copied from a pattern elsewhere?

Reply:
I updated the docstring a bit in https://github.com/numpy/numpy/pull/19459/commits/38b0fc25b4bc054d399ecfe54e30596888cda21e, but the general gist of it is to avoid direct usage of the `__init__` constructor.

Directly assigning the exception attributes, as is done by `_reconstruct`, allows us to avoid having to deal with its overload.
In addition, it also removes the need to extract the (potential) message prefix from the exception message.

It is in principle possible to get rid of it, but at the cost of more complexity in `__reduce__`.

Reply:
> It is in principle possible to get rid of it, but at the cost of more complexity in `__reduce__`.

As an example:

``` python
def __reduce__(self):
    cls = type(self)
    axis = self.axis
    ndim = self.ndim
    message = self.args[0]

    if axis is ndim is None:
        return cls, (message,)
    else:
        prefix, *_ = message.partition(":")
        return cls, (axis, ndim, prefix or None)
      
```